src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_search.cs

```diff
@@ -717,19 +717,51 @@ public bool IsNext(ReadOnlySpan<T> next, bool advancePast = false)
         [MethodImpl(MethodImplOptions.NoInlining)]
         private unsafe bool IsNextSlow(ReadOnlySpan<T> next, bool advancePast)
         {
-            // Call PeekSlow directly since we know there isn't enough space.
-            Debug.Assert(UnreadSpan.Length < next.Length);
+            ReadOnlySpan<T> currentSpan = UnreadSpan;
 
-            T* t = stackalloc T[next.Length];
-            ReadOnlySpan<T> peek = PeekSlow(new Span<T>(t, next.Length));
+            // We should only come in here if we need more data than we have in our current span
+            Debug.Assert(currentSpan.Length < next.Length);
 
-            if (next.SequenceEqual(peek))
+            int length = next.Length;
+            SequencePosition nextPosition = _nextPosition;
+
+            while (true)
             {
-                if (advancePast)
+                if (!next.StartsWith(currentSpan))
                 {
-                    Advance(next.Length);
+                    break;
+                }
+
+                if (next.Length == currentSpan.Length)
+                {
+                    // Fully matched
+                    if (advancePast)
+                    {
+                        Advance(length);
+                    }
+                    return true;
+                }
+                next = next.Slice(currentSpan.Length);
+
+                // Need to check the next segment
+                while (true)
+                {
+                    if (!Sequence.TryGet(ref nextPosition, out ReadOnlyMemory<T> nextSegment, advance: true))
+                    {
+                        // Nothing left
+                        return false;
+                    }
+
+                    if (nextSegment.Length > 0)
+                    {
+                        currentSpan = nextSegment.Span;
+                        if (currentSpan.Length > next.Length)
+                        {
+                            currentSpan = currentSpan.Slice(0, next.Length);
```

Comment:
This slice and then checking if next.Length == currentSpan.Length within the outerloop on success seems unnecessary.

Instead of slicing, why not have the outerloop break when next.StartsWith(currentSpan) returns true.

Reply:
If we have `"ABCD"` left in `currentSpan`, it won't start `next` if it has `"AB"` left in it, but it should match.